pkg/blobstore/cassandra/cassandra_blob_access.go

```diff
@@ -0,0 +1,673 @@
+		prometheus.MustRegister(segmentReadHist)
+		prometheus.MustRegister(totalWritesHist)
+		prometheus.MustRegister(lastAccessUpdatesFailures)
+	})
+
+	cluster := gocql.NewCluster(clusterHosts...)
+	cluster.Keyspace = keyspace
+	cluster.Consistency = writeConsistency
+	if preferredDc != "" {
+		cluster.PoolConfig.HostSelectionPolicy = gocql.TokenAwareHostPolicy(gocql.DCAwareRoundRobinPolicy(preferredDc))
+	}
+
+	if port != 0 {
+		cluster.Port = int(port)
+	}
+
+	if protoVersion != 0 {
+		cluster.ProtoVersion = int(protoVersion)
+	}
+
+	if username != "" {
+		cluster.Authenticator = gocql.PasswordAuthenticator{
+			Username: username,
+			Password: password,
+		}
+	}
+
+	if tlsConfig != nil {
+		cluster.SslOpts = &gocql.SslOptions{
+			Config: tlsConfig,
+		}
+	} else {
+		log.Printf("Not setting up TLS config")
+	}
+
+	if tablePrefix == "" {
+		log.Printf("No table prefix set. Cowardly refusing to proceed.")
+		return blobstore.NewErrorBlobAccess(errors.New("table prefix must be set"))
+	}
+	log.Printf("Table prefix: %s", tablePrefix)
+
+	log.Printf("Creating session in keyspace '%s' for %v", keyspace, clusterHosts)
+
+	session, err := cluster.CreateSession()
+	if err != nil {
+		log.Printf("Unable to create session: %v", err)
+		return blobstore.NewErrorBlobAccess(err)
+	}
+
+	// Create the worker pool for updating the last access times
+	jobs := make(chan job, lastAccessUpdateWorkerCount)
+	for i := 0; i < lastAccessUpdateWorkerCount; i++ {
+		go lastUpdateWorker(jobs)
+	}
+
+	log.Printf("Cassandra storage ready.")
+
+	return &cassandraBlobAccess{
+		capabilitiesProvider:     capabilitiesProvider,
+		session:                  session,
+		readBufferFactory:        readBufferFactory,
+		segmentSize:              segmentSizeInBytes,
+		lastAccessUpdateInterval: lastAccessUpdateInterval,
+		lastAccessUpdate:         jobs,
+		universalInstanceName:    universalInstanceName,
+		tables:                   newTables(tablePrefix, session),
+	}
+}
+
+func (a *cassandraBlobAccess) GetCapabilities(ctx context.Context, instanceName bbdigest.InstanceName) (*remoteexecution.ServerCapabilities, error) {
+	return a.capabilitiesProvider.GetCapabilities(ctx, instanceName)
+}
+
+func logErrorIfNotCancelledContext(err error, msg string, args ...interface{}) {
+	if err != nil && !errors.Is(err, context.Canceled) {
+		initialMessage := fmt.Sprintf(msg, args...)
+		log.Printf("%s: %v", initialMessage, err)
+	}
+}
+
+func (a *cassandraBlobAccess) Get(ctx context.Context, digest bbdigest.Digest) buffer.Buffer {
+	metadata, err := a.tables.metadata.read(ctx, digest)
+	if err != nil {
+		if errors.Is(err, gocql.ErrNotFound) {
+			return buffer.NewBufferFromError(status.Errorf(codes.NotFound, "No metadata found for digest %s", digest.String()))
+		}
+		return buffer.NewBufferFromError(err)
+	}
+
+	go a.tables.metadata.updateLastAccessTime(context.Background(), digest, metadata.instanceName, metadata.lastAccess, a.lastAccessUpdateInterval, a.lastAccessUpdate)
```

Comment:
Again, even though having no concurrency limits in place might at this point never have been an issue for you, I'd like to see this addressed before this change lands.